crates/CrateAssembler.kt

```diff
@@ -157,12 +168,40 @@ class CrateAssembler : Callable<Unit> {
             cargoToml.set<Config>("lib", this)
             set<String>("path", crateRootPath)
         }
+
+        val canonicalDeps = universeManifests.flatMap {
+            TomlParser().parse(it.inputStream())
+                    .getOrElse("dependencies", Config.inMemory()).entrySet().asSequence()
+        }.associate { it.key to it.getValue<String>() }
+
+        val bazelDepFeatures = depFeaturesList.associate { it.split("=").let { (dep, feats) -> dep to feats } }
+                .mapValues { (_, feats) -> feats.split(",") }
+
         cargoToml.createSubConfig().apply {
             cargoToml.set<Config>("dependencies", this)
             depsList.associate { it.split("=").let { (dep, ver) -> dep to ver } }.forEach { (dep, ver) ->
-                set<String>(dep, ver)
+                if (canonicalDeps.contains(dep))
+                    set<String>(dep, canonicalDeps[dep])
+                else if (bazelDepFeatures.containsKey(dep)) {
+                    val depConfig = Config.inMemory()
+                    set<Config>(dep, depConfig)
+                    depConfig.set<String>("version", "=$ver")
+                    depConfig.set("features", bazelDepFeatures[dep])
+                } else
+                    set<String>(dep, "=$ver")
```

Comment:
Note: we _are_ going against the grain here by pinning the version in this way and not allowing it to vary. This can become a problem for the end user, especially if they are depending on another crate with this quirk: they would have to compile e.g. uuid v1.1.2, v1.1.3, and v1.1.4, because various libraries expect the exact versions of all, rather than being able to just depend on the highest SemVer-compatible version. This bloats the size of the binary, and greatly increases the compile and link times.

Reply:
Note: discussion with the rest of the engineering team concludes that pinning the exact version all the way down is desirable for us, since we prioritise repeatable builds over user applications' compile times & a bit of size.